lib/ssz.ex

```diff
@@ -139,4 +164,349 @@ defmodule Ssz do
+  end
+
+  def serialize(value) when is_binary(value), do: {:ok, value}
+
+  def serialize(value, []) when is_binary(value), do: {:ok, value}
+
+  def serialize(value, %{type: :struct}) do
+    serialize(value)
+  end
+
+  def serialize(elements, %{type: :list, schema: schema, max_size: max_size}) do
+    if length(elements) > max_size do
+      {:error, "max_size_error"}
+    else
+      serialize_list(elements, schema)
+    end
+  end
+
+  def serialize(value, %{type: :uint, size: size}) when is_integer(value),
+    do: serialize_uint(value, size)
+
+  def serialize(value, %{type: :bytes, size: _size}), do: {:ok, value}
+
+  def serialize(value, %{} = schema),
+    do: {:error, "Unknown schema: #{inspect(schema)} for value #{inspect(value)}"}
+
+  @spec serialize_list(list, list) :: {:ok, binary} | {:error, String.t()}
+  def serialize_list(elements, schema) when is_list(elements) do
+    fixed_parts =
+      elements
+      |> Enum.map(fn element ->
+        if is_variable_size(schema), do: nil, else: serialize(element, schema)
+      end)
+      |> Enum.map(fn
+        {:ok, ser} -> ser
+        nil -> nil
+      end)
+
+    variable_parts =
+      elements
+      |> Enum.map(fn element ->
+        if is_variable_size(schema), do: serialize(element, schema), else: <<>>
+      end)
+      |> Enum.map(fn
+        {:ok, ser} -> ser
+        <<>> -> <<>>
+      end)
+
+    fixed_lengths =
+      fixed_parts
+      |> Enum.map(fn part ->
+        if part != nil, do: byte_size(part), else: @bytes_per_length_offset
+      end)
+
+    variable_lengths =
+      variable_parts
+      |> Enum.map(fn part -> byte_size(part) end)
+
+    if Enum.sum(fixed_lengths ++ variable_lengths) <
+         2 ** (@bytes_per_length_offset * @bits_per_byte) do
+      variable_offsets =
+        0..(length(elements) - 1)
+        |> Enum.map(fn i ->
+          slice_variable_lengths = Enum.take(variable_lengths, i)
+          sum = Enum.sum(fixed_lengths ++ slice_variable_lengths)
+          serialize_uint(sum, 32)
+        end)
+        |> Enum.map(fn {:ok, ser} -> ser end)
+
+      fixed_parts = get_fixed_parts(fixed_parts, variable_offsets)
+
+      final_ssz =
+        (fixed_parts ++ variable_parts)
+        |> :binary.list_to_bin()
+
+      {:ok, final_ssz}
+    else
+      {:error, "invalid lengths"}
+    end
+  end
+
+  @spec get_fixed_parts(list, list) :: list
+  defp get_fixed_parts(fixed_parts, variable_offsets) do
+    fixed_parts
+    |> Enum.with_index()
+    |> Enum.map(fn {part, index} ->
+      if part != nil, do: part, else: Enum.at(variable_offsets, index)
+    end)
+  end
+
+  defp serialize_uint(value, size) do
+    <<encoded::binary-size(div(size, 8)), _rest::binary>> =
+      value
+      |> :binary.encode_unsigned(:little)
+      |> String.pad_trailing(div(size, 8), <<0>>)
+
+    {:ok, encoded}
+  end
+
+  ### DECODE ######
+  @spec decode_elixir(binary, module) :: {:ok, struct} | {:error, String.t()}
+  def decode_elixir(bin, schema) do
```

Comment:
We should rename this to `deserialize`, to be consistent in the naming, and avoid confusion with Elixir structs.